### PR TITLE
21519 - Rework partner disbursement job so it works with multiple link and unlinking

### PR DESCRIPTION
**Pull request by seeker25**
*Issue #:*
https://github.com/bcgov/entity/issues/21519

*Description of changes:*
Requires some changes on the pay-queue as well to handle the new identifiers etc
Handle invoice refunds 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-pay license (Apache 2.0).

### CODE REVIEW

File: jobs/payment-jobs/tasks/ap_task.py [hunk 1]
```diff
@@ -105,13 +107,51 @@ def _create_routing_slip_refund_file(cls):  # pylint:disable=too-many-locals, to
 
             cls._create_file_and_upload(ap_content)
 
+    @staticmethod
+    def get_invoices_for_disbursement(partner):
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/ap_task.py]: We're going to have to rewrite these at a later date

File: pay-api/src/pay_api/models/partner_disbursements.py [hunk 2]
```diff
@@ -38,27 +38,23 @@ class PartnerDisbursements(BaseModel):  # pylint: disable=too-many-instance-attr
...
-    disbursement_type = db.Column('disbursement_type', db.String(50), nullable=False)
     feedback_on = db.Column('feedback_on', db.DateTime, nullable=True)
     partner_code = db.Column('partner_code', db.String(50), nullable=False)
     processed_on = db.Column('processed_on', db.DateTime, nullable=True)
     is_reversal = db.Column('is_reversal', db.Boolean(), nullable=False, default=False)
-    source_gl = db.Column('source_gl', db.String(50), nullable=False)
     status_code = db.Column('status_code', db.String(25), nullable=False)
     target_id = db.Column(db.Integer, nullable=True)
-    target_gl = db.Column('target_gl', db.String(50), nullable=False)
+    target_type = db.Column(db.String(50), nullable=True)
```
Comment (seeker25) [inline pay-api/src/pay_api/models/partner_disbursements.py:60]: need to update migration

File: pay-api/src/pay_api/services/eft_short_names.py [hunk 3]
```diff
@@ -220,6 +220,18 @@ def _reverse_payment_action(cls, short_name_id: int, statement_id: int):  # pyli
         # pick up on this and set to REFUNDED when processed for reversal. Invoice disbursement status will be
         # the final indicator on whether this is complete
 
+        """
+        Need to create row and update existing rows for:
```
Comment (seeker25) [inline pay-api/src/pay_api/services/eft_short_names.py:224]: Ody has this already, removing

File: jobs/payment-jobs/tasks/ejv_partner_distribution_task.py [hunk 4]
```diff
@@ -53,308 +54,221 @@ def create_ejv_file(cls):
...
-        invoices: List[InvoiceModel] = db.session.query(InvoiceModel) \
-            .filter(InvoiceModel.invoice_status_code == InvoiceStatus.PAID.value) \
-            .filter(
-            InvoiceModel.payment_method_code.notin_([PaymentMethod.INTERNAL.value,
-                                                     PaymentMethod.DRAWDOWN.value,
-                                                     PaymentMethod.EFT.value])) \
-            .filter((InvoiceModel.disbursement_status_code.is_(None)) |
-                    (InvoiceModel.disbursement_status_code == DisbursementStatus.ERRORED.value)) \
-            .filter(~InvoiceModel.receipts.any(cast(ReceiptModel.receipt_date, Date) >= disbursement_date.date())) \
+        base_query = db.session.query(InvoiceModel, PaymentLineItemModel, DistributionCodeModel) \
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/ejv_partner_distribution_task.py:67]: I think this is potentially too many columns, we'll see

File: jobs/payment-jobs/tasks/ejv_partner_distribution_task.py [hunk 5]
```diff
@@ -53,308 +54,221 @@ def create_ejv_file(cls):
...
+            .join(DistributionCodeModel,
+                  DistributionCodeModel.distribution_code_id == PaymentLineItemModel.fee_distribution_id) \
+            .filter(PartnerDisbursementsModel.status_code.is_(None)) \
+            .filter(PartnerDisbursementsModel.corp_type_code == partner.code) \
+            .filter(DistributionCodeModel.stop_ejv.is_(False) | DistributionCodeModel.stop_ejv.is_(None)) \
+            .order_by(DistributionCodeModel.distribution_code_id, PaymentLineItemModel.id) \
+            .all()
+
+        for partner_disbursement, distribution_code in partner_disbursements:
+            suffix = 'PR' if partner_disbursement.disbursement_type == EJVLinkType.PARTIAL_REFUND else 'PD'
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/ejv_partner_distribution_task.py:124]: Wanted something different to show up on the flow_through or description_identifier

File: jobs/payment-jobs/tasks/ejv_partner_distribution_task.py [hunk 6]
```diff
@@ -53,308 +54,221 @@ def create_ejv_file(cls):
...
+            disbursement_rows.append(Disbursement(
+                bcreg_distribution_code=distribution_code,
+                partner_distribution_code=distribution_code.partner_distribution_code,
+                line_item=DisbursementLineItem(
+                    amount=partner_disbursement.amount,
+                    flow_through=f'{suffix}{partner_disbursement.target_id:<110}',
+                    description_identifier=f'{suffix}#{partner_disbursement.id}',
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/ejv_partner_distribution_task.py:131]: 130 + 131 need some cleanup and thought again

File: jobs/payment-jobs/tasks/ejv_partner_distribution_task.py [hunk 7]
```diff
@@ -53,308 +54,221 @@ def create_ejv_file(cls):
...
-            partner_distribution_code_ids: List[int] = db.session.scalars(query.filter(
-                DistributionCodeModel.client == bc_reg_client_code
-            )).all()
-        else:
-            # Rule for GI. Debit is 112 and credit is not 112.
-            partner_distribution_code_ids: List[int] = db.session.scalars(query.filter(
-                DistributionCodeModel.client != bc_reg_client_code
-            )).all()
+            .filter(DistributionCodeModel.disbursement_distribution_code_id.is_(None)) \
+            .filter_conditionally(True if batch_type == 'GA' else None,
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/ejv_partner_distribution_task.py:257]: work around for `filter_conditionally`

might just make another `filter_boolean`

File: pay-api/src/pay_api/services/eft_service.py [hunk 8]
```diff
@@ -65,6 +67,16 @@ def create_invoice(self, payment_account: PaymentAccount, line_items: List[Payme
                        **kwargs) -> None:
         """Do nothing here, we create invoice references on the create CFS_INVOICES job."""
         self.ensure_no_payment_blockers(payment_account)
+        if corp_type := CorpTypeModel.find_by_code(invoice.corp_type_code):
+            if corp_type.has_partner_disbursement:
+                PartnerDisbursementsModel(
```
Comment (seeker25) [inline pay-api/src/pay_api/services/eft_service.py:72]: neeed to update this

File: pay-api/src/pay_api/services/eft_service.py [hunk 9]
```diff
@@ -142,21 +154,3 @@ def _create_refund_model(cls, request: Dict[str, str],
         )
         refund.status = EFTCreditInvoiceStatus.PENDING_REFUND
         return refund
-
-    @classmethod
-    def _render_email_body(cls, shortname: str, amount: str, comment: str, shortname_id: str) -> str:
```
Comment (seeker25) [inline pay-api/src/pay_api/services/eft_service.py:147]: Moved this to email service as we discussed @Jxio

File: jobs/payment-jobs/tasks/ejv_partner_distribution_task.py [hunk 10]
```diff
@@ -53,308 +54,221 @@ def create_ejv_file(cls):
         cls._create_ejv_file_for_partner(batch_type='GI')  # Internal ministry
         cls._create_ejv_file_for_partner(batch_type='GA')  # External ministry
 
-    # TODO grab EFT invoices from PartnerDisbursements, eventually everything will run through partner disbursements.
     @staticmethod
-    def get_invoices_for_disbursement(partner):
+    def get_disbursement_by_distribution_for_partner(partner) -> List[Disbursement]:
         """Return invoices for disbursement. Used by EJV and AP."""
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/ejv_partner_distribution_task.py:59]: comment fix

File: jobs/payment-jobs/tasks/common/dataclasses.py [hunk 11]
```diff
@@ -12,15 +12,40 @@
...
+    description_identifier: str
+    is_reversal: bool
+    is_legacy: bool
+    disbursement_type: str
+    identifier: int
+
+
+@dataclass
+class Disbursement:
+    """Data class to hold disbursement line item."""
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/common/dataclasses.py:42]: comment

File: pay-api/migrations/versions/2024_07_18_aae01971bd53_.py [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+Create Date: 2024-07-18 10:51:20.058891
+
+"""
+from alembic import op
+import sqlalchemy as sa
+from sqlalchemy.dialects import postgresql
+
+# revision identifiers, used by Alembic.
+revision = 'aae01971bd53'
+down_revision = 'fb59bf68146d'
```
Comment (Jxio) [inline pay-api/migrations/versions/2024_07_18_aae01971bd53_.py:14]: this down_revision need to update, I had another commit to migration yesterday

**Comment by seeker25** [comment]
will reopen

**Comment by seeker25** [comment]
Missing pay queue changes, so it knows when to set the partner disbursements status

**Comment by seeker25** [comment]
Waiting for CAS before merging, so I can try this on the test environment

File: jobs/payment-jobs/tasks/common/dataclasses.py [hunk 13]
```diff
@@ -12,15 +12,40 @@
...
+    identifier: int
+
+
+@dataclass
+class Disbursement:
+    """DTO mapping for disbursement."""
+
+    bcreg_distribution_code: DistributionCodeModel
+    partner_distribution_code: DistributionCodeModel
+    line_item: DisbursementLineItem
```
Comment (ochiu) [inline jobs/payment-jobs/tasks/common/dataclasses.py:46]: yay for dataclasses

File: jobs/payment-jobs/tasks/ejv_partner_distribution_task.py [hunk 14]
```diff
@@ -54,230 +56,226 @@ def create_ejv_file(cls):
         cls._create_ejv_file_for_partner(batch_type='GA')  # External ministry
 
     @staticmethod
-    def get_invoices_for_disbursement(partner):
-        """Return invoices for disbursement. Used by EJV and AP."""
+    def get_disbursement_by_distribution_for_partner(partner) -> List[Disbursement]:
+        """Return disbursements dataclass for partners."""
+        # Internal invoices aren't disbursed to partners, DRAWDOWN is handled by the mainframe.
+        # EFT is handled by the PartnerDisbursements table.
+        # ##################################################### Original (Legacy way) - invoice.disbursement_status_code
```
Comment (ochiu) [inline jobs/payment-jobs/tasks/ejv_partner_distribution_task.py:63]: Thanks for this comment

File: jobs/payment-jobs/tasks/ejv_partner_distribution_task.py [hunk 15]
```diff
@@ -54,230 +56,226 @@ def create_ejv_file(cls):
...
+                  DistributionCodeModel.distribution_code_id == PaymentLineItemModel.fee_distribution_id) \
+            .filter(PartnerDisbursementsModel.status_code == DisbursementStatus.WAITING_FOR_JOB.value) \
+            .filter(PartnerDisbursementsModel.partner_code == partner.code) \
+            .filter(DistributionCodeModel.stop_ejv.is_(False) | DistributionCodeModel.stop_ejv.is_(None)) \
+            .filter(~InvoiceModel.receipts.any(cast(ReceiptModel.receipt_date, Date) >= disbursement_date.date())) \
+            .order_by(DistributionCodeModel.distribution_code_id, PaymentLineItemModel.id) \
+            .all()
+
+        for partner_disbursement, payment_line_item, distribution_code in partner_disbursements:
+            suffix = 'PR' if partner_disbursement.target_type == EJVLinkType.PARTIAL_REFUND else ''
```
Comment (ochiu) [inline jobs/payment-jobs/tasks/ejv_partner_distribution_task.py:129]: EJVLinkType.PARTIAL_REFUND.value?

**Review by ochiu — APPROVED** [review]

File: jobs/payment-jobs/tasks/ejv_partner_distribution_task.py (hunk 15 above)
Comment (seeker25) [reply jobs/payment-jobs/tasks/ejv_partner_distribution_task.py:129]: Not implemented, will implement partial refunds in the ticket

**Comment by ochiu** [comment]
I appreciate the comments outlining the current state and anticipated future state for the Partner Disbursements table

**Review by Jxio — APPROVED** [review]
Looks good to me!

**Comment by seeker25** [comment]
Left to go:

1. Migration for existing EFT invoices
2. Side by side testing with old code
3. I think the invoice.total is incorrect.. we need to disburse excluding service fees

File: pay-api/src/pay_api/services/eft_refund.py [hunk 16]
```diff
@@ -114,19 +117,19 @@ def handle_invoice_refund(invoice: InvoiceModel,
...
-                #         amount=reversal_total,
-                #         is_reversal=True,
-                #         partner_code=invoice.corp_type_code,
-                #         status_code=DisbursementStatus.WAITING_FOR_JOB.value,
-                #         target_id=invoice.id,
-                #         target_type=EJVLinkType.INVOICE.value
-                #     ).flush()
+                if reversal_total > 0:
+                    PartnerDisbursementsModel(
+                        amount=reversal_total,
```
Comment (seeker25) [inline pay-api/src/pay_api/services/eft_refund.py:126]: I don't think this is right

File: pay-api/src/pay_api/services/eft_service.py [hunk 17]
```diff
@@ -80,6 +82,16 @@ def create_invoice(self, payment_account: PaymentAccount, line_items: List[Payme
                        **kwargs) -> None:
         """Do nothing here, we create invoice references on the create CFS_INVOICES job."""
         self.ensure_no_payment_blockers(payment_account)
+        if corp_type := CorpTypeModel.find_by_code(invoice.corp_type_code):
+            if corp_type.has_partner_disbursements:
+                PartnerDisbursementsModel(
+                    amount=invoice.total,
```
Comment (seeker25) [inline pay-api/src/pay_api/services/eft_service.py:88]: not right

File: pay-api/src/pay_api/services/eft_service.py [hunk 18]
```diff
@@ -226,20 +238,20 @@ def reverse_payment_action(short_name_id: int, statement_id: int):
...
-        #         target_type=EJVLinkType.INVOICE.value
-        #     ).flush()
+            if corp_type := CorpTypeModel.find_by_code(invoice.corp_type_code):
+                if corp_type.has_partner_disbursements and current_link.amount > 0:
+                    invoice_disbursements.setdefault(invoice, 0)
+                    invoice_disbursements[invoice] += current_link.amount
+
+        for invoice, total_amount in invoice_disbursements.items():
+            PartnerDisbursementsModel(
+                amount=total_amount,
```
Comment (seeker25) [inline pay-api/src/pay_api/services/eft_service.py:248]: not right, includes service fees

**Comment by seeker25** [comment]
Just trying EFT + EFT reversals locally now, after that I'll merge

**Pull request merged**